N24DCCN167/BaiTap_buoi8/7.cpp

```diff
@@ -0,0 +1,35 @@
+#include <iostream>
+#include <vector>
+using namespace std;
+int main()
+{
+    string str = "    Cau  lac    bo   ITMC    ";
+    string newStr = "";
+    vector<string> words;
+    for (int i = 0; i < str.length(); i++)
+    {
+        if (str[i] != ' ')
+        {
+            newStr += str[i];
+        }
+        else
+        {
+            if (newStr != "")
+            {
+                words.push_back(newStr);
+                newStr = "";
+            }
+        }
+    }
+    if (str[str.length() - 1] != ' ')
+        words.push_back(newStr);
+    string res = "";
+    for (int i = 0; i < words.size(); i++)
+    {
+        res = res + words[i];
+        if (i != words.size() - 1)
+            res += ' ';
+    }
+    cout << "'" << res << "'";
+    return 0;
```

Comment:
ý tưởng của em có thể sẽ là tách từng từ đưa vào mảng, sau đó in lại mảng đó thêm 1 dấu space ở giửa đúng k? Cách này chạy dc. Nhưng sẽ tốn bộ nhớ (do lưu thêm 1 mảng) và tốn thời gian (do có 2 for rời). Cách tối ưu hơn là sử dụng while kết hợp với hàm find() để tìm dấu space dư, sau đó xóa nó đi. while chạy cho tới khi nào hết space dư. Các đó sẽ tối ưu hơn. Vì câu này gặp rất nhiều trong các đề thi, nên thuật toán ở đây phải tối ưu nhất có thể. Em update lại nha